verilog/tools/extractor/verilog_extractor.cc

```diff
@@ -0,0 +1,332 @@
+                                                                          value_(leaf.get().text()) {};
+
+    void print() {
+        std::cout <<
+                  "{\n"
+                  "\"StartLocation\": " << startLocation_ << ",\n" <<
+                  R"("EndLocation": )" << endLocation_ << ",\n" <<
+                  R"("Value": ")" << value_ << "\"\n" <<
+                  "}"
+                  << std::endl;
+    }
+
+private:
+    int startLocation_, endLocation_;
+    std::string value_;
+};
+
+class Block {
+public:
+    Block(Type datatype = Type::NoType) : dataType_(datatype) {}
+
+    Block(std::vector<Anchor> anchor, Type dataType) :
+            anchors_(std::move(anchor)), dataType_(dataType) {}
+
+    void print() {
+        std::cout << "{\n";
+
+        if (dataType_ != Type::NoType) {
+            std::cout << R"("Type": ")" << dataType_ << "\",\n";
+        }
+
+        std::cout << "\"Anchor\": [\n";
+        for (int i = 0; i < anchors_.size(); i++) {
+            anchors_[i].print();
+            if (i != anchors_.size() - 1) {
+                std::cout << ",\n";
+            }
+        }
+        std::cout << "],\n";
+
+        std::cout << "\"Children\": [\n";
+        for (int i = 0; i < children_.size(); i++) {
+            children_[i].print();
+            if (i != children_.size() - 1) {
+                std::cout << ",\n";
+            }
+        }
+        std::cout << "]\n";
+
+        std::cout << "}\n";
+    }
+
+    std::vector<Block> Children() { return children_; };
+
+    void AppendChild(Block entry) { children_.push_back(std::move(entry)); };
+
+    void AppendChild(std::vector<Block> children) {
+        children_.insert(children_.end(), children.begin(), children.end());
+    };
+
+    void AppendAnchor(Anchor entry) { anchors_.push_back(std::move(entry)); };
+
+    void AppendAnchor(std::vector<Anchor> anchors) {
+        anchors_.insert(anchors_.end(), anchors.begin(), anchors.end());
+    };
+private:
+    std::vector<Anchor> anchors_;
+
+    Type dataType_;
+
+    std::vector<Block> children_;
+};
+
+using TagExtractor = std::function<Block(const verible::SyntaxTreeNode &, const verible::SyntaxTreeNode &,
+                                         absl::string_view)>;
+std::map<int, TagExtractor> tagExtractor;
+
+std::vector<Block> Extract(const verible::SyntaxTreeLeaf &node, absl::string_view base) {
+    std::cout << "Start Leaf" << std::endl;
+    std::cout << verilog::NodeEnumToString(static_cast<verilog::NodeEnum>(node.Tag().tag)) << " <<>> " << node.Tag().tag
+              << " " << node.get()
+              << std::endl;
+    std::cout << "End Leaf" << std::endl;
+    std::cout << std::endl;
+    std::vector<Block> blocks;
+    return blocks;
+}
+
+std::vector<Block> Extract(const verible::SyntaxTreeNode &node, absl::string_view base) {
+    std::cout << "Start Node" << std::endl;
```

Comment:
instead printing to std::cout you could log some parts of the code for debug purpose. That log lines can be submitted.

Example:
LOG(INFO) << "Start leaf " << ...;
See: https://github.com/google/verible/blob/master/common/util/logging.h

Running the binary with --alsologtostderr will print the log statements to your terminal (otherwise they only go to a file)

Reply:
@fangism can you provide how this is used in Verible?

Reply:
In a lot of places I use:

```
VLOG(N) << ...;
```

for N ranging up to 4 (super verbose).